### PR TITLE
[KeyVault] adding soft-delete support for KeyVault certificates

### PR DESCRIPTION
**Pull request by dragav**
## Description

---

This checklist is used to make sure that common guidelines for a pull request are followed. You can find a more complete discussion of PowerShell cmdlet best practices [here](https://msdn.microsoft.com/en-us/library/dd878270(v=vs.85).aspx).

- [ ] **I have read the [contribution guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md).**
- [ ] **If changes were made to any cmdlet, the XML help was regenerated using the [platyPSHelp module](https://github.com/Azure/azure-powershell/blob/preview/documentation/help-generation.md).**
- [ ] **If any large changes are made to a service, they are reflected in the respective [change log](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#updating-the-change-log).**

### [General Guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#general-guidelines)
- [ ] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/preview/documentation/cleaning-up-commits.md).
- [ ] The pull request does not introduce [breaking changes](https://github.com/Azure/azure-powershell/blob/preview/documentation/breaking-changes/breaking-changes-definition.md) (unless a major version change occurs in the assembly and module).

### [Testing Guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#testing-guidelines)
- [ ] Pull request includes test coverage for the included changes.
- [ ] PowerShell scripts used in tests should do any necessary setup as part of the test or suite setup, and should not use hard-coded values for locations or existing resources.

### [Cmdlet Signature Guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#cmdlet-signature-guidelines)
- [ ] New cmdlets that make changes or have side effects should implement `ShouldProcess` and have `SupportShouldProcess=true` specified in the cmdlet attribute. You can find more information on `ShouldProcess` [here](https://github.com/Azure/azure-powershell/wiki/PowerShell-Cmdlet-Design-Guidelines#supportsshouldprocess).
- [ ] Cmdlet specifies `OutputType` attribute if any output is produced - if the cmdlet produces no output, it should implement a `PassThru` parameter.

### [Cmdlet Parameter Guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#cmdlet-parameter-guidelines)
- [ ] Parameter types should not expose types from the management library - complex parameter types should be defined in the module.
- [ ] Complex parameter types are discouraged - a parameter type should be simple types as often as possible. If complex types are used, they should be shallow and easily creatable from a constructor or another cmdlet.
- [ ] Cmdlet parameter sets should be mutually exclusive - each parameter set must have at least one mandatory parameter not in other parameter sets.

### CODE REVIEW

**Comment by dragav** [comment]
The build appears to be failing because of static analysis; there is a large number of warnings, none of which come from the KeyVault namespace. 

The only references to KeyVault listed in the build warnings file are several skipped management test cases.

File: src/ResourceManager/KeyVault/Commands.KeyVault.Test/Scripts/VaultCertificateTests.ps1 [hunk 1]
```diff
@@ -614,4 +614,142 @@ function Test_UpdateCertificateTags
...
+    $certName = Get-CertificateName $rootCertName + '1' 
+    $createdCert1 = CreateAKVCertificate $keyVault $certName
+    Assert-NotNull $createdCert1
+
+    $certName = Get-CertificateName $rootCertName + '2'
+    $createdCert2 = CreateAKVCertificate $keyVault $certName
+    Assert-NotNull $createdCert2
+
+    Get-AzureKeyVaultCertificate $keyVault |  Where-Object {$_.CertificateName -like $rootCertName + '*'}  | Remove-AzureKeyVaultCertificate -Force -Confirm:$false
+    Wait-Seconds 30
```
Comment (cormacpayne) [inline src/ResourceManager/KeyVault/Commands.KeyVault.Test/Scripts/VaultCertificateTests.ps1]: @dragav why do we need to wait 30 seconds?

File: src/ResourceManager/KeyVault/Commands.KeyVault.Test/app.config [hunk 2]
```diff
@@ -0,0 +1,11 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (cormacpayne) [inline src/ResourceManager/KeyVault/Commands.KeyVault.Test/app.config]: @dragav please remove this file

File: src/ResourceManager/KeyVault/Commands.KeyVault/Commands/GetAzureKeyVaultCertificate.cs [hunk 3]
```diff
@@ -146,5 +171,20 @@ private void GetAndWriteCertificatesVersions(string vaultName, string name, stri
...
+        {
+            KeyVaultObjectFilterOptions options = new KeyVaultObjectFilterOptions
+            {
+                VaultName = VaultName,
+                NextLink = null
+            };
+
+            do
+            {
+                var pageResults = DataServiceClient.GetDeletedCertificates(options);
```
Comment (cormacpayne) [inline src/ResourceManager/KeyVault/Commands.KeyVault/Commands/GetAzureKeyVaultCertificate.cs]: @dragav can we change this block to something like:

```cs
var pageResults = new List<DeletedCertificateIdentityItem>()
do
{
    pageResults.Add( DataServiceClient.GetDeletedCertificates( options ) );
} while ( !string.IsNullOrEmpty( options.NextLink ) );

WriteObject( pageResults, true );
```

this will ensure that all of the results are enumerated as a collection at once

File: src/ResourceManager/KeyVault/Commands.KeyVault/Commands/UndoAzureKeyVaultCertificateRemoval.cs [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+
+using System.Management.Automation;
+using Microsoft.Azure.Commands.KeyVault.Models;
+using Microsoft.Azure.KeyVault.Models;
+
+namespace Microsoft.Azure.Commands.KeyVault
+{
+    [Cmdlet( VerbsCommon.Undo, "AzureKeyVaultCertificateRemoval",
+    SupportsShouldProcess = true,
+    ConfirmImpact = ConfirmImpact.Low,
```
Comment (cormacpayne) [inline src/ResourceManager/KeyVault/Commands.KeyVault/Commands/UndoAzureKeyVaultCertificateRemoval.cs]: @dragav why change the `ConfirmImpact` level?

File: src/ResourceManager/KeyVault/Commands.KeyVault/help/AzureRM.KeyVault.md [hunk 5]
```diff
@@ -20,9 +20,6 @@ Adds a contact for certificate notifications.
 ### [Add-AzureKeyVaultKey](Add-AzureKeyVaultKey.md)
 Creates a key in a key vault or imports a key into a key vault.
 
-### [Add-AzureKeyVaultManagedStorageAccount](Add-AzureKeyVaultManagedStorageAccount.md)
```
Comment (cormacpayne) [inline src/ResourceManager/KeyVault/Commands.KeyVault/help/AzureRM.KeyVault.md]: @dragav it seems like this file did not get refreshed properly

File: src/ResourceManager/KeyVault/Commands.KeyVault/Models/KeyAttributes.cs [hunk 6]
```diff
@@ -65,7 +65,7 @@ public KeyAttributes()
 
         public DateTime? Updated { get; private set; }
 
-        public bool PurgeDisabled { get; private set; }
+        public string RecoveryLevel { get; private set; }
```
Comment (cormacpayne) [inline src/ResourceManager/KeyVault/Commands.KeyVault/Models/KeyAttributes.cs]: @dragav removing the `PurgeDisabled` property is a breaking change. Users who run the following will be broken:

```powershell
$recoveredKey = Undo-AzureKeyVaultKeyRemoval -VaultName $vault -Name $keyName
$recoveredKey.Attributes.PurgeDisabled
```

Our recommendation is to keep supporting this property, making it obsolete until the next breaking change release (this fall) when you can then remove it since users will have been warned ahead of time of the change.

File: src/ResourceManager/KeyVault/Commands.KeyVault/Models/KeyIdentityItem.cs [hunk 7]
```diff
@@ -36,7 +36,7 @@ internal KeyIdentityItem(Azure.KeyVault.Models.KeyItem keyItem, VaultUriHelper v
             NotBefore = keyItem.Attributes.NotBefore;
             Created = keyItem.Attributes.Created;
             Updated = keyItem.Attributes.Updated;
-            PurgeDisabled = keyItem.Attributes.PurgeDisabled;
```
Comment (cormacpayne) [inline src/ResourceManager/KeyVault/Commands.KeyVault/Models/KeyIdentityItem.cs]: @dragav same comment throughout this file

File: src/ResourceManager/KeyVault/Commands.KeyVault/Commands/RemoveAzureKeyVaultCertificate.cs [hunk 8]
```diff
@@ -27,7 +27,7 @@ namespace Microsoft.Azure.Commands.KeyVault
         SupportsShouldProcess = true,
         ConfirmImpact = ConfirmImpact.High,
         HelpUri = Constants.KeyVaultHelpUri)]
-    [OutputType(typeof(KeyVaultCertificate))]
+    [OutputType(typeof(DeletedCertificateBundle))]
```
Comment (cormacpayne) [inline src/ResourceManager/KeyVault/Commands.KeyVault/Commands/RemoveAzureKeyVaultCertificate.cs]: @dragav nothing has changed about the output of this cmdlet, right? It is still returning a `KeyVaultCertificate` below the block of code you added. The block you added does not write any object to the output stream, either.

File: src/ResourceManager/KeyVault/Commands.KeyVault/help/Add-AzureKeyVaultKey.md [hunk 9]
```diff
@@ -12,176 +12,36 @@ Creates a key in a key vault or imports a key into a key vault.
 
 ## SYNTAX
 
-### Create (Default)
```
Comment (dragav) [inline src/ResourceManager/KeyVault/Commands.KeyVault/help/Add-AzureKeyVaultKey.md]: @cormacpayne help generation seems to output incorrect content for a few commands, I haven't figured out why. I've manually updated some of them, and this one slipped through; will revert the change.

File: src/ResourceManager/KeyVault/Commands.KeyVault.Test/Scripts/VaultCertificateTests.ps1 (hunk 1 above)
Comment (dragav) [inline src/ResourceManager/KeyVault/Commands.KeyVault.Test/Scripts/VaultCertificateTests.ps1]: Soft-deletion is an asynchronous operation, performed in the background by a different role.

File: src/ResourceManager/KeyVault/Commands.KeyVault/Commands/GetAzureKeyVaultCertificate.cs (hunk 3 above)
Comment (dragav) [inline src/ResourceManager/KeyVault/Commands.KeyVault/Commands/GetAzureKeyVaultCertificate.cs]: This follows the existing pattern we use for keys and secrets. The reason we print this one page at a time is to be responsive. This operation could take a long time (and even fail/time out) when the number of versions is high.

File: src/ResourceManager/KeyVault/Commands.KeyVault/Models/KeyAttributes.cs (hunk 6 above)
Comment (dragav) [inline src/ResourceManager/KeyVault/Commands.KeyVault/Models/KeyAttributes.cs]: Normally I'd heed your recommendation; in this case, however, the presence of this field has security implications. The server protocol does not include this field, and a default here conveys the wrong information. I would rather have the client be broken (and make a quick script change) than use incorrect information.

Furthermore, the previous PSH release was based on a non-public KV SDK; that is, this field should have been considered "pre-release" via PSH as well, but we did not find a way to mark it as such. This PSH release uses a public KV SDK, and so it is in line with the REST API spec and the service itself.

I can add this as an entry to the breaking change log..

File: src/ResourceManager/KeyVault/Commands.KeyVault/Commands/RemoveAzureKeyVaultCertificate.cs (hunk 8 above)
Comment (dragav) [inline src/ResourceManager/KeyVault/Commands.KeyVault/Commands/RemoveAzureKeyVaultCertificate.cs]: Not sure what you mean.. The response of this cmdlet is instantiated on line 93:

DeletedCertificateBundle certBundle = null;

The DataServiceClient (invoked to delete the cert) returns:

 DeletedCertificateBundle DeleteCertificate(string vaultName, string certName);

Comment (cormacpayne) [inline src/ResourceManager/KeyVault/Commands.KeyVault/Commands/RemoveAzureKeyVaultCertificate.cs]: @dragav the object being written to the output stream is still a `KeyVaultCertificate` on line 111, so the output type has not changed

File: src/ResourceManager/KeyVault/Commands.KeyVault/Models/KeyAttributes.cs (hunk 6 above)
Comment (markcowl) [inline src/ResourceManager/KeyVault/Commands.KeyVault/Models/KeyAttributes.cs]: We don't have the option of the client being broken, we will not ship any breaking changes until November, except in the case of critical security issues or other critical bugs.

Comment (dragav) [inline src/ResourceManager/KeyVault/Commands.KeyVault/Models/KeyAttributes.cs]: This can't be a breaking change, since PurgeDisabled was never part of the service REST API spec. Yes, there is a publicly-released PowerShell out there which includes PurgeDisabled, but nobody should have taken a dependency on that since the feature was in preview.

Comment (markcowl) [inline src/ResourceManager/KeyVault/Commands.KeyVault/Models/KeyAttributes.cs]: Unfortunately, no breaking changes are a hard requirement

Comment (dragav) [inline src/ResourceManager/KeyVault/Commands.KeyVault/Models/KeyAttributes.cs]: Understood; reverted the breaking change.

**Comment by cormacpayne** [comment]
On-demand: http://azuresdkci.cloudapp.net/view/1-AzurePowerShell/job/powershell-demand/1826/

**Comment by markcowl** [comment]
@azuresdkci retest this please

File: src/ResourceManager/KeyVault/ChangeLog.md [hunk 10]
```diff
@@ -24,6 +24,10 @@
 ## Version 3.3.0
 
 ## Version 3.2.1
+* New/updated Cmdlets to support soft-delete for KeyVault certificates
```
Comment (cormacpayne) [inline src/ResourceManager/KeyVault/ChangeLog.md]: @dragav please move this to under `Current Release`; the versions that are already listed are the notes from previous releases
